### PR TITLE
[Merged by Bors] - Bugfix for: Druid connection not established

### LINKED ISSUE #174: Druid connection is not established if Druid is started afterwards
If the DruidConnection is applied before the Druid cluster is started then the DruidConnection remains in status "Pending". The reconciliation should also happen when the Druid cluster changes.

### PR DESCRIPTION
**Pull request by fhennig**
## Description

Fixes #174 

A test for the bug is found in this PR https://github.com/stackabletech/integration-tests/pull/235

- the DruidConnection controller now watches the ConfigMaps, to be notified if a druid instance that should be connected is deployed; to connect it then.
- the DruidConnnection namespaces are also made optional in this PR

## Review Checklist
- [x] Code contains useful comments
- [x] (Integration-)Test cases added (or not applicable)
- [x] Documentation added (or not applicable)
- [x] Changelog updated (or not applicable)
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
- [x] Helm chart can be installed and deployed operator works (or not applicable)

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/crd/src/druidconnection.rs [hunk 1]
```diff
@@ -36,6 +36,34 @@ impl DruidConnection {
...
+    }
+
+    pub fn superset_namespace(&self) -> String {
+        if let Some(ns) = &self.spec.superset.namespace {
+            ns.to_string()
+        } else if let Some(ns) = &self.namespace() {
+            ns.to_string()
+        } else {
+            "default".to_string()
+        }
```
Comment (maltesander) [inline rust/crd/src/druidconnection.rs:51]: I think this should be an error. If neither of the 2 matches above there is something seriously wrong :) I would not opt for default then.

Or it just returns an Option<String> and "None" for the last case (seeing how it is used later) and let the framework deal with it.

File: rust/operator-binary/src/main.rs [hunk 2]
```diff
@@ -179,9 +180,9 @@ async fn main() -> anyhow::Result<()> {
                             .state()
                             .into_iter()
                             .filter(move |druid_connection| {
-                                &druid_connection.spec.superset.namespace
+                                &druid_connection.superset_namespace()
```
Comment (maltesander) [inline rust/operator-binary/src/main.rs:183]: No need for unwraps here if superset_namespace returns an Option.

**Review by maltesander — COMMENTED** [review]
Just one thing (not sure if that applies, but i think it could make things easier)

File: rust/operator-binary/src/main.rs (hunk 2 above)
Comment (maltesander) [reply rust/operator-binary/src/main.rs:183]: Edit: I messed up the lines here, i meant the code (comparison) below.

**Review by siegfriedweber — APPROVED** [review]
Approved as soon as the open comments of @maltesander are done.

File: rust/crd/src/druidconnection.rs (hunk 1 above)
Comment (maltesander) [reply rust/crd/src/druidconnection.rs:51]: I think the panic would be fine here. Just setting default and ignoring a bigger problem will complicate finding the problem.

**Review by maltesander — COMMENTED** [review]
Some small things:

File: rust/operator-binary/src/main.rs [hunk 3]
```diff
@@ -179,10 +180,13 @@ async fn main() -> anyhow::Result<()> {
...
-                                    == sdb.metadata.namespace.as_ref().unwrap()
-                                    && &druid_connection.spec.superset.name
-                                        == sdb.metadata.name.as_ref().unwrap()
+                                if let Ok(ns) = druid_connection.superset_namespace() {
+                                    &ns == sdb.metadata.namespace.as_ref().unwrap()
+                                        && &druid_connection.superset_name()
+                                            == sdb.metadata.name.as_ref().unwrap()
+                                } else {
+                                    false
+                                }
```
[the remark this answers is not in the document]
Comment (maltesander) [reply rust/operator-binary/src/main.rs:189]: Or even shorter
```suggestion
                                sdb.metadata.namespace == druid_connection.superset_namespace().ok()
                                    && Some(druid_connection.superset_name()) == sdb.metadata.name
```

File: rust/operator-binary/src/druid_connection_controller.rs [hunk 4]
```diff
@@ -89,6 +93,7 @@ impl ReconcilerError for Error {
             Error::GetImportJob { import_job, .. } => Some(import_job.clone().erase()),
             Error::DruidDiscoveryCheck { .. } => None,
             Error::SupersetDBRetrieval { .. } => None,
+            Error::DruidConnectionNoNamespace { .. } => None,
```
[the remark this answers is not in the document]
Comment (fhennig) [reply rust/operator-binary/src/druid_connection_controller.rs]: done

File: rust/operator-binary/src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (fhennig) [reply rust/operator-binary/src/main.rs:189]: a, good idea

**Review by maltesander — APPROVED** [review]
LGTM.

**Comment by fhennig** [comment]
bors merge

**Pull request closed without merging**